### ISSUE TITLE
Gateway restores fail on Ignition 8 when volume mounted

### ISSUE THREAD

**Issue by ptrdlbrg** [thirdgen88/ignition-docker#8]
**TL;DR**: When a docker volume is mounted at `/var/lib/ignition` then gateway restores fail on Ignition 8 containers.  This can be fixed by making sure the Ignition temp directory is also under `/var/lib/ignition`.

**Specifics:**
When a docker volume is mounted at `/var/lib/ignition` then gateway restores fail with the following error on Ignition 8 containers:
```
com.inductiveautomation.ignition.gateway.localdb.DBStartupException: Error restoring internal database from backup.
jvm 2    |      at com.inductiveautomation.ignition.gateway.localdb.LocalDBManagerImpl.setup(LocalDBManagerImpl.java:162)
jvm 2    |      at com.inductiveautomation.ignition.gateway.IgnitionGateway.startupInternal(IgnitionGateway.java:820)
jvm 2    |      at com.inductiveautomation.ignition.gateway.redundancy.RedundancyManagerImpl.startup(RedundancyManagerImpl.java:272)
jvm 2    |      at com.inductiveautomation.ignition.gateway.IgnitionGateway.initRedundancy(IgnitionGateway.java:664)
jvm 2    |      at com.inductiveautomation.ignition.gateway.IgnitionGateway.lambda$initInternal$0(IgnitionGateway.java:602)
jvm 2    |      at com.inductiveautomation.ignition.common.execution.impl.BasicExecutionEngine$ThrowableCatchingRunnable.run(BasicExecutionEngine.java:518)
```

I've found that Ignition gateway restore operation fails if the Ignition temp directory is on a different file system than the Ignition data directory. A docker volume is mounted as a different file system than rest of the container file system, and thus it can trigger this problem. 

This problem of crossing file systems is almost certainly an Ignition bug, but probably is rarely seen in customer configurations. 

Anyway, this problem pops up in the Ignition 8 containers but not the Ignition 7 containers. This is because the temp directory is under /var/lib/ignition in the Ignition 7 containers but isn't in Ignition 8 containers. The Ignition installer sets the temp directory to /var/lib/ignition/temp but since the Ignition 8 containers are created using the zip file instead of the installer, you have to manually set the temp directory to the right location. 

The fix for this is changing the Ignition 8 Dockerfile from having this:
```
# Stage data and user-lib in var folder
RUN mkdir -p /var/lib/ignition && \
    mv data /var/lib/ignition/ && \
    mv user-lib /var/lib/ignition/ && \
    ln -s /var/lib/ignition/data data && \
    ln -s /var/lib/ignition/user-lib user-lib
```
to this
```
# Stage data and user-lib in var folder
RUN mkdir -p /var/lib/ignition && \
    mv data /var/lib/ignition/ && \
    mv user-lib /var/lib/ignition/ && \
    mv temp /var/lib/ignition/ && \
    ln -s /var/lib/ignition/data data && \
    ln -s /var/lib/ignition/user-lib user-lib  && \
    ln -s /var/lib/ignition/temp temp
```

**Comment by thirdgen88** [comment]
Great catch @ptrdlbrg.  Unfortunately, my recent work to support third-party modules has driven me to suggest that folks target their named volume at `/var/lib/ignition/data` instead of the full `/var/lib/ignition` path.  This is done so that when containers are upgraded from one version to the next, only the `data` is preserved and the modules from `user-lib` are still bound to the container and thus are changed out with new ones in the upgraded container image.  I was able to reproduce your failure with a volume at `/var/lib/ignition/data`.  I may try linking temp from `/usr/local/share/ignition/temp` to `/var/lib/ignition/data/temp` and see if that could get us going.  I don’t like storing that temp data in with the user volume, but we may have little choice until the underlying issue is addressed.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by thirdgen88** [comment]
I went ahead and mapped the `temp` directory into `/var/lib/ignition/data/temp` and was able to successfully restore without an error with a named volume bound to `/var/lib/ignition/data` (as is now the recommendation).  The fix is in e55d11907b64c067c3d659f070b3232c9e9ef986.  I’ve triggered a build and Docker Hub should have the updated images.  @ptrdlbrg, can you pull the latest and confirm that it is working for you?

**Comment by ptrdlbrg** [comment]
Whether or not a user would include `user-lib` in their volume depends on their use case, so I like how you've structured it so you can include it or exclude it based on which directory you set your volume to. Your solution of putting `temp` inside of the `data` directory is a reasonable solution.

I agree that I would prefer to keep `temp` out of the volume, but unfortunately that doesn't seem to be an option right now. 

I pulled the latest 8.0.2 and confirmed that restores now work as they should.  Thanks for the quick response. Thank you for creating and maintaining these docker images. I really appreciate it!

**Comment by ptrdlbrg** [comment]
I [reported the underlying bug](https://forum.inductiveautomation.com/t/bug-14227-restoring-gateway-backup-fails-data-and-temp-are-on-different-linux-file-systems/25927) on the Ignition forum.
